includes/class-shortcodes.php

```diff
@@ -0,0 +1,486 @@
+<?php
+namespace Activitypub;
+
+class Shortcodes {
+	/**
+	 * The post object we're currently working on
+	 *
+	 * @var WP_Post  $post    A WordPress Post Object
+	 */
+	private $post;
+
+	/**
+	 * Class constructor, registering WordPress then shortcodes
+	 *
+	 * @param WP_Post  $post    A WordPress Post Object
+	 */
+	public function __construct( $post = null ) {
+		if( $post == null ) {
+			$post = \get_post();
+		}
+
+		if( \is_object( $post ) && ! $post instanceof WP_POST ) {
+			$this->post = $post;
+		} else {
+			$this->post = false;
+		}
+
+		foreach( get_class_methods( $this ) as $shortcode ) {
+			if( strpos( $shortcode, '__' ) !== 0 ) {
+				add_shortcode( 'ap_' . $shortcode, array( $this, $shortcode ) );
+			}
+		}
+	}
+
+	/**
+	 * Generates output for the ap_hashtags shortcode
+	 *
+	 * @param array  $atts 		shortcode attributes
+	 * @param string $content 	shortcode content
+	 * @param string $tag 		shortcode tag name
+	 *
+	 * @return string
+	 */
+	public function hashtags( $atts, $content, $tag ) {
+		if( $this->post === false ) {
+			return '';
+		}
+
+		$tags = \get_the_tags( $this->post->ID );
+
+		if ( ! $tags ) {
+			return '';
+		}
+
+		$hash_tags = array();
+
+		foreach ( $tags as $tag ) {
+			$hash_tags[] = \sprintf( '<a rel="tag" class="u-tag u-category" href="%s">#%s</a>', \get_tag_link( $tag ), $tag->slug );
+		}
+
+		return \implode( ' ', $hash_tags );
+	}
+
+	/**
+	 * Generates output for the ap_title shortcode
+	 *
+	 * @param array  $atts 		shortcode attributes
+	 * @param string $content 	shortcode content
+	 * @param string $tag 		shortcode tag name
+	 *
+	 * @return string
+	 */
+	public function title( $atts, $content, $tag ) {
+		if( $this->post === false ) {
+			return '';
+		}
+
+		return \get_the_title( $this->post->ID );;
+	}
+
+	/**
+	 * Generates output for the ap_excerpt shortcode
+	 *
+	 * @param array  $atts 		shortcode attributes
+	 * @param string $content 	shortcode content
+	 * @param string $tag 		shortcode tag name
+	 *
+	 * @return string
+	 */
+	public function excerpt( $atts, $content, $tag ) {
+		if( $this->post === false ) {
+			return '';
+		}
+
+		$length = ACTIVITYPUB_EXCERPT_LENGTH;
```

Comment:
use `shortcode_atts` to set default params: https://developer.wordpress.org/reference/functions/shortcode_atts/

Reply:
Done.